crypto/envelope.go

```diff
@@ -0,0 +1,139 @@
+package crypto
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"github.com/golang/protobuf/proto"
+	pb "github.com/libp2p/go-libp2p-core/crypto/pb"
+)
+
+// SignedEnvelope contains an arbitrary []byte payload, signed by a libp2p peer.
+// Envelopes are signed in the context of a particular "domain", which is a string
+// specified when creating and verifying the envelope. You must know the domain
+// string used to produce the envelope in order to verify the signature and
+// access the contents.
+type SignedEnvelope struct {
+
+	// The public key that can be used to verify the signature and derive the peer id of the signer.
+	PublicKey PubKey
+
+	// A binary identifier that indicates what kind of data is contained in the payload.
+	// TODO(yusef): enforce multicodec prefix
+	TypeHint []byte
+
+	// The envelope payload. This is private to discourage accessing the payload without verifying the signature.
+	// To access, use the Open method.
+	contents []byte
+
+	// The signature of the domain string, type hint, and contents.
+	signature []byte
+}
+
+var errEmptyDomain = errors.New("envelope domain must not be empty")
+
+// MakeEnvelope constructs a new SignedEnvelope using the given privateKey.
+//
+// The required 'domain' string contextualizes the envelope for a particular purpose,
+// and must be supplied when verifying the signature.
+//
+// The 'typeHint' field indicates what kind of data is contained and may be empty.
+func MakeEnvelope(privateKey PrivKey, domain string, typeHint []byte, contents []byte) (*SignedEnvelope, error) {
+	if len(domain) == 0 {
+		return nil, errEmptyDomain
+	}
+	toSign := makeSigBuffer(domain, typeHint, contents)
+	sig, err := privateKey.Sign(toSign)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SignedEnvelope{
+		PublicKey: privateKey.GetPublic(),
+		TypeHint:  typeHint,
+		contents:  contents,
+		signature: sig,
+	}, nil
+}
+
+// UnmarshalEnvelope converts a serialized protobuf representation of an envelope
+// into a SignedEnvelope struct.
+func UnmarshalEnvelope(serializedEnvelope []byte) (*SignedEnvelope, error) {
```

Comment:
Two step signature verification processes where the user can forget the second step make me nervous. How about:

* `UnmarshalWithoutValidating(bytes) (*SignedEnvelope, error)`
* `Open(bytes, domain) (*SignedEnvelope, error)`

That way, we can't end up with an unvalidated `SignedEnvelope` object unless the user very explicitly opts in.

Reply:
I like this - I tried to force validation by making the payload private & requiring you to "open" the envelope (and validate) to get the payload. But you could still end up passing on an invalid envelope to someone else if you never open the payload.

Should we also be validating the signature before we marshal to bytes? Then if someone invalidates the signature by changing e.g. the `payloadType` field, they won't be able to write the record out. We'd need the `Marshal` method to take the domain string in order to validate. It might also be worth having an unsafe `MarshalWithoutValidating` method for testing or if you explicitly want to opt out of validation.

Reply:
> Should we also be validating the signature before we marshal to bytes?

Meh. I'm fine with marshaling without verifying. We could also just make all the fields private but that may not be worth it. I don't really have any strong opinions here.

I also didn't realize that the payload was private. That resolves my initial concern (although the type is really a part of the payload and should probably also be private). However, I think validating when deserializing is likely going to be more efficient in general and make it easier to pass around signed envelopes.

Reply:
Hmm, maybe instead of validating before marshaling, we can just make all the fields private, and expose read-only accessor methods, so it's not possible to alter the struct contents after you make or open an envelope?

Reply:
after sync discussion with @vyzo & @Stebalien we decided to just expose the fields - if you mutate them it's your own fault when validation fails 😄